### PR TITLE
More ray differentials work

### PR DESCRIPTION
**Pull request by est77**
- Add setters to Dual class.
- Use duals instead of vectors in places where differentials are relevant.
- Do not use BSDFSample::get_incoming() in BSDFs.

### CODE REVIEW

File: src/appleseed/foundation/math/dual.h [hunk 1]
```diff
@@ -102,6 +104,13 @@ inline const T& Dual<T>::get_value() const
 }
 
 template <typename T>
+inline void Dual<T>::set_value(const T& value)
+{
+    m_value = value;
+    m_has_derivatives = false;
```
Comment (dictoon) [inline src/appleseed/foundation/math/dual.h]: That's dangerous, e.g. if `set_value()` is called after `set_derivatives()`. I would recommend against clearing the flag here.

File: src/appleseed/foundation/math/dual.h [hunk 2]
```diff
@@ -123,6 +132,14 @@ const T& Dual<T>::get_dy() const
     return m_dy;
 }
 
+template <typename T>
+void Dual<T>::set_derivatives(const T& dx, const T& dy)
```
Comment (dictoon) [inline src/appleseed/foundation/math/dual.h]: Mark as inline.

File: src/appleseed/renderer/kernel/lighting/pathtracer.h [hunk 3]
```diff
@@ -341,14 +348,31 @@ size_t PathTracer<PathVisitor, Adjoint>::trace(
...
+            sample.get_incoming_vector(),
             ray.m_time,
             ray.m_dtime,
             bsdf_mode_to_ray_flags(sample.get_mode()),
             ray.m_depth + 1);
 
+        if (sample.get_incoming().has_derivatives())
+        {
+            scattered_ray.m_rx.m_org =
+                    scattered_ray.m_org + vertex.m_shading_point->get_dpdx();
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/pathtracer.h]: Fix indentation.

File: src/appleseed/renderer/modeling/bsdf/oslashikhminbrdf.cpp [hunk 4]
```diff
@@ -148,20 +147,22 @@ namespace
             const double sin_theta = sqrt(1.0 - cos_theta * cos_theta);
 
             // Compute the halfway vector in world space.
-            h = sample.get_shading_basis().transform_to_parent(
+            const Vector3d h = sample.get_shading_basis().transform_to_parent(
                         Vector3d::unit_vector(cos_theta, sin_theta, cos_phi, sin_phi));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/oslashikhminbrdf.cpp]: Fix indentation.

**Comment by dictoon** [comment]
Looking pretty good!

File: src/appleseed/foundation/math/dual.h (hunk 1 above)
Comment (est77) [inline src/appleseed/foundation/math/dual.h]: It is expected that you call set_value() before set_derivatives().
But checking the code, we don't need setters at all and then there's no 
possibility of doing the wrong thing anymore.

**Pull request merged**